### PR TITLE
TEPHRA-101 Add SnapshotCodecV3 to enable CDAP migration

### PR DESCRIPTION
**Pull request by ghelmling**

### CODE REVIEW

**Comment by sreevatsanraman** [comment]
LGTM :+1:

File: tephra-core/src/main/java/co/cask/tephra/TransactionManager.java [hunk 1]
```diff
@@ -490,13 +490,20 @@ private void restoreSnapshot(TransactionSnapshot snapshot) {
...
-      if (entry.getValue().getExpiration() < 0) {
-        long writePointer = entry.getKey();
-        long expiration = getTxExpirationFromWritePointer(writePointer, defaultLongTimeout);
+      long writePointer = entry.getKey();
+      long expiration = entry.getValue().getExpiration();
+      // LONG transactions will either have a negative expiration or expiration set to the long timeout
+      // use a fudge factor on the expiration check, since expiraton is set based on system time, not the write pointer
+      if (entry.getValue().getType() == null &&
+          (expiration < 0 ||
+              (getTxExpirationFromWritePointer(writePointer, defaultLongTimeout) - expiration < 10000))) {
```
Comment (nitinmotgi) [inline tephra-core/src/main/java/co/cask/tephra/TransactionManager.java]: Should ```10000``` be a constant defined somewhere or should it be a config ?

Comment (ghelmling) [inline tephra-core/src/main/java/co/cask/tephra/TransactionManager.java]: I'm not sure it should be either.  If we really want to expose a way to tweak this without a code change, we can make it configurable.  But I don't think it should be documented.  This entire method should be removed as soon as possible -- as soon as we can drop support for the old codec.

**Comment by albertshau** [comment]
lgtm

**Pull request merged**